### DISCUSSION TITLE
session.commit() goes into hung state for an hour when RDS reboot with failover is peformed

### DISCUSSION THREAD

**Discussion by roshanpathak500** [sqlalchemy/sqlalchemy#12167] [Usage Questions]
We are using sqlalchemy (version 2.0.36)  as the ORM framework for our python (version 3.8) application using _create_engine_ method from _sqlalchemy_ and _sessionmaker_, _scoped_session_ methods from _sqlalchemy.orm_

The way we create engine is as shown below 

```
engine = create_engine(DATABASE_URI, 
                       pool_size=30, 
                       max_overflow=20, 
                       pool_pre_ping=True, 
                       pool_recycle=60*60
```

The session is created as shown below 

`session = scoped_session(sessionmaker(autocommit=False, autoflush=True, bind=engine))`

Using this session object , we use `session.commit()` snippet to perform DB queries 

The issue is whenever there is a RDS reboot with failover , the code gets stuck on session.commit() for an hour without throwing "SSL SYSCALL error: Operation timed out" exception . Not just this, the hung state blocks the main process on the EKS pod in which python application is running in turn not allowing any requests to be honored . 

When I simply shutdown the instance , the process immediately fails .

I have tried sqlalchemy's pessimistic and optimistic failure strategies but none of them seem to be working .

Adding the below to create_engine arguments does not help either

                    
```
connect_args={
                            "connect_timeout": 10*60, 
                            "options": "-c statement_timeout=300000",
                             "keepalives": 1, 
                            "keepalives_idle": 30,
                            "keepalives_interval": 5,
                            "keepalives_count": 5,
                           }
                        )
```

We have another python application deployed in the same EKS cluster in a sibling pod leveraging _twisted.enterprise.adbapi_ for connection pooling strategy and it immediately fails with "SSL SYSCALL error: Operation timed out" during a reboot with failover scenario. 

What should I do in the sqlalchemy based python application to ensure the error is immediately thrown ? 

The solution of this thread did not help 
https://groups.google.com/g/sqlalchemy/c/NANPH1QWgMM

**Comment by zzzeek** [comment]
this is a database driver issue.  what database driver are you using?

**Reply by sajanav** [reply]
PostgreSQL 13.15

**Reply by zzzeek** [reply]
not a driver.   

drivers are:

[psycopg2](https://www.psycopg.org/docs/)

[pg8000](https://pypi.org/project/pg8000/)

[asyncpg](https://github.com/MagicStack/asyncpg)

etc

**Reply by roshanpathak500** [reply]
[psycopg2](https://www.psycopg.org/docs/)

**Reply by roshanpathak500** [reply]
The DATABASE_URI looks like this  "postgresql+psycopg2://dummyuser:dummypassword@myhost:5432/myappdb"
The only psycopg2 dependency we have in our application is _psycopg2-binary_ of version 2.9.3.

What could have possibly gone wrong with psycopg2 ?

**Reply by zzzeek** [reply]
if failover is taking the current IP number and removing it from routing, it may be becoming a black hole.   you'd want to experiment with the psycopg2 driver directly; make a connection to the RDS database, do the failover, then try to emit further commands on that connection.  See if they time out or just hang indefinitely.  

here's an article, by the Guardian's own tech team, where they seem to have the identical issue that you're having

https://www.theguardian.com/info/2019/dec/02/faster-postgresql-connection-recovery

> What happened is that all the TCP (Transmission Control Protocol) connections of the pool got into a sort of “zombie” state: broken yet not closed. The reason for that is the postgreSQL driver does not set a TCP timeout on the connections. The pool never attempted to close the connections and create new ones, yet no data would come in or out of these connections. 

it looks like there's a sockettimeout option they mention, but psycopg2 doesnt seem to have this as it's based on libpq.

it also seems like the statement_timeout you're using, which [psycopg2 has an open bug that this does not work with SSL](https://github.com/psycopg/psycopg2/issues/944)

then i dont seem to be able to get any further...possibly try adding [tcp_user_timeout](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS) not clear

you can probably improve the situation using [NullPool](https://docs.sqlalchemy.org/en/20/core/pooling.html#switching-pool-implementations) however this still wont help connections that are in current transactions, if TCP is going into a zombie state

**Reply by roshanpathak500** [reply]
Have already tried with making the connection directly with psycopg2 , code snippet below .
**Issue still  persists** 

```
from psycopg2.pool import SimpleConnectionPool

def get_psycopg2_conn():
        try:
            connection_pool = SimpleConnectionPool(
                minconn=5,
                maxconn=30,
                keepalives=1,
                keepalives_idle=30,
                keepalives_interval=10,
                keepalives_count=5,
                host=“myhost”,
                database=“myappdb”,
                user=“dummyuser”,
                password=“dummypassword”
            )
            return connection_pool.getconn()
        except:
            return None
```

`raw_connection = get_psycopg2_conn()`

```
engine2 = create_engine(
    DATABASE_URI,
    creator=lambda: raw_connection  
)
```

`session = scoped_session(sessionmaker(autocommit=False, autoflush=True, bind=engine2))`

Have also tried [tcp_user_timeout](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS) directly in the DATABASE_URI string ; it had not worked either . 

NULLPOOL by documentation did not seem like helping with the current situation , hence did not use it .

**Reply by zzzeek** [reply]
we're trying to see if not using connection pooling at all can get you around this issue, since idle connections are most likely to be affected by failovers like this

**Reply by sajanav** [reply]
```
async_engine = create_async_engine(db_url_async, pool_recycle=600)
async_session = async_sessionmaker(async_engine, expire_on_commit=False)

...
    async with async_session() as session:
        
        await session.execute(
...
        )
        await session.commit() 
```
so this answer to a similar thread :https://github.com/sqlalchemy/sqlalchemy/discussions/10903 will it help ?

**Reply by roshanpathak500** [reply]
@zzzeek Tried the below as suggested without any pooling mechanism in place
```
 conn = psycopg2.connect(
                host="myhost"
                database="myappdb",
                user="dummyuser",
                password=""dummypassword"            
                )
```

The flow is still stuck in session.commit() when the _RDS reboot with failover_ hit .
